File: scratch/check_syntax.py
import sys

def check_braces(filename):
    with open(filename, 'r', encoding='utf-8') as f:
        content = f.read()
    
    # Extract contents of <script> tag
    start_tag = '<script>'
    end_tag = '</script>'
    
    start_pos = content.find(start_tag)
    while start_pos != -1:
        end_pos = content.find(end_tag, start_pos)
        if end_pos == -1:
            print(f"Error: Missing {end_tag}")
            return
        
        script_content = content[start_pos + len(start_tag):end_pos]
        
        stack = []
        for i, char in enumerate(script_content):
            if char == '{':
                stack.append(('{', i))
            elif char == '}':
                if not stack:
                    print(f"Error: Unmatched '}}' at script index {i}")
                    # Print context
                    start = max(0, i-20)
                    end = min(len(script_content), i+20)
                    print(f"Context: ...{script_content[start:end]}...")
                    return
                stack.pop()
        
        if stack:
            char, i = stack.pop()
            print(f"Error: Unmatched '{{' at script index {i}")
            start = max(0, i-20)
            end = min(len(script_content), i+20)
            print(f"Context: ...{script_content[start:end]}...")
            return
        
        print("Done checking one script tag. No brace mismatch found.")
        start_pos = content.find(start_tag, end_pos)

if __name__ == "__main__":
    check_braces(sys.argv[1])

File: scratch/check_syntax.js
const fs = require('fs');

function checkBraces(filename) {
    const content = fs.readFileSync(filename, 'utf8');
    const startTag = '<script>';
    const endTag = '</script>';

    let startPos = content.indexOf(startTag);
    while (startPos !== -1) {
        const endPos = content.indexOf(endTag, startPos);
        if (endPos === -1) {
            console.log(`Error: Missing ${endTag}`);
            return;
        }

        const scriptContent = content.substring(startPos + startTag.length, endPos);
        const stack = [];
        let inString = null;
        let isEscaped = false;

        for (let i = 0; i < scriptContent.length; i++) {
            const char = scriptContent[i];
            
            if (isEscaped) {
                isEscaped = false;
                continue;
            }
            if (char === '\\') {
                isEscaped = true;
                continue;
            }

            if (inString) {
                if (char === inString) inString = null;
                continue;
            }

            if (char === "'" || char === '"' || char === '`') {
                inString = char;
                continue;
            }

            if (char === '{') {
                stack.push({ char, i });
            } else if (char === '}') {
                if (stack.length === 0) {
                    console.log(`Error: Unmatched '}' at script index ${i}`);
                    console.log(`Context: ...${scriptContent.substring(Math.max(0, i - 40), Math.min(scriptContent.length, i + 40))}...`);
                    return;
                }
                stack.pop();
            }
        }

        if (stack.length > 0) {
            const last = stack.pop();
            console.log(`Error: Unmatched '{' at script index ${last.i}`);
            console.log(`Context: ...${scriptContent.substring(Math.max(0, last.i - 40), Math.min(scriptContent.length, last.i + 40))}...`);
            return;
        }

        console.log("Done checking one script tag. No brace mismatch found.");
        startPos = content.indexOf(startTag, endPos);
    }
}

checkBraces(process.argv[2]);
